ecommerce/api.py

```diff
@@ -103,3 +111,184 @@ def make_reference_id(order):
     return (
         f"{_REFERENCE_NUMBER_PREFIX}{settings.CYBERSOURCE_REFERENCE_PREFIX}-{order.id}"
     )
+
+
+def get_valid_coupon_versions(coupons, user, auto_only=False):
+    """
+    Given a list of coupon ids, determine which of them are valid based on invoice version dates and redemptions.
+
+    Args:
+        coupons (QuerySet of coupon ids): List of coupons to filter for validity
+        auto_only (bool): Whether or not to filter by automatic=True
+
+    Returns:
+        list of CouponVersion: CouponVersion objects sorted by discount, highest first.
+    """
+    valid_coupons = []
+    now = now_in_utc()
+
+    # Get the ids of the latest coupon versions
+    cv_latest = (
+        CouponVersion.objects.select_related()
+        .filter(coupon__in=coupons)
+        .order_by("coupon", "-created_on")
+        .distinct("coupon")
+        .values_list("pk")
+    )
+
+    # filter by expiration and activation dates
+    query = (
+        CouponVersion.objects.select_related()
+        .filter(pk__in=cv_latest)
+        .filter(
+            Q(invoice_version__expiration_date__gte=now)
+            | Q(invoice_version__expiration_date__isnull=True)
+        )
+        .filter(
+            Q(invoice_version__activation_date__lte=now)
+            | Q(invoice_version__activation_date__isnull=True)
+        )
+    )
+
+    if auto_only:
+        query = query.filter(invoice_version__automatic=True)
+
+    # filter by redemption counts
+    for cv in query:
+        redemptions_global = CouponRedemption.objects.filter(coupon_version=cv)
+        redemptions_user = redemptions_global.filter(order__purchaser=user)
+        if (
```

Comment:
Can we look into doing this filtering at the database level? Or is that too unwieldy?

Reply:
That's what I had hoped to do, but it started to get unwieldy due to having to compare counts.

Reply:
Sounds good